File: 0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def calArea(self,nums):
        stack = []
        result = float("-inf")
        for i in range(len(nums)) :
            while stack and nums[i] <= nums[stack[-1]]:
                ind = stack.pop()
                pse = stack[-1] if stack else -1
                result = max(result,(i - pse - 1)*nums[ind])
            stack.append(i)
        while stack :
            ind = stack.pop()
            pse = stack[-1] if stack else -1
            result = max(result, (len(nums)-pse-1)*nums[ind])

        return result

    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        nums = []
        n,m = len(matrix) , len(matrix[0])
        result = float("-inf")
        for i in range(n):
            if i == 0 :
                for j in matrix[0]:
                    nums.append(int(j))
            else :
                for j in range(m) :
                    if matrix[i][j]=='0':
                        nums[j] = 0
                    else :
                        nums[j]+=1
            result = max(result,self.calArea(nums))
        return result
            



        

File: 0085-maximal-rectangle/0085-maximal-rectangle.java
class Solution {
    public int calMaxArea(char[][] matrix,int row,int n){
        int[] nums = new int[n];
        int pse, ind ,result = Integer.MIN_VALUE;
        for(int i =0;i<= row;i++){
            if(i==0){
                for(int j=0;j<n;j++){
                    if(matrix[i][j]=='1')
                        nums[j] = 1;
                }
            }
            else{
                for(int j=0;j<n;j++){
                    if(matrix[i][j]== '1')
                        nums[j]+= 1;
                    else 
                        nums[j] =0;
                }
            }
        }
        Stack<Integer> stack = new Stack<>();
        for(int i=0;i<n;i++){
            while(! stack.isEmpty() && nums[i] <= nums[stack.peek()]){
                ind = stack.pop();
                pse = stack.isEmpty() ? -1 : stack.peek();
                result = Math.max(result , (i - pse-1)*nums[ind]);

            }
            stack.push(i);
        }
        while(!stack.isEmpty()){
            ind = stack.pop();
            pse = stack.isEmpty()? -1 : stack.peek();
            result = Math.max(result,nums[ind]*(n - pse-1));

        }
        return result ;

    }
    public int maximalRectangle(char[][] matrix) {
        int m = matrix.length ;
        int n = matrix[0].length ;
        int result = Integer.MIN_VALUE;
        for(int i =0;i<m;i++){
            result = Math.max(result,calMaxArea(matrix,i,n));

        }
        return result ;


        
    }
}